tutorials/dataframe/df107_SingleTopAnalysis.py

```diff
@@ -0,0 +1,245 @@
+parser.add_argument("-b", action="store_true", default=False, help="Use ROOT batch mode")
+parser.add_argument("-t", action="store_true", default=False, help="Use implicit multi threading (running always on the full dataset)")
+args = parser.parse_args()
+if args.b: ROOT.gROOT.SetBatch(True)
+if args.t:
+    ROOT.EnableImplicitMT()
+    lumi_scale = 1.0
+else:
+    lumi_scale = args.lumi_scale
+
+# Create a ROOT dataframe for each dataset
+# Note that we load the filenames from an external json file.
+path = "root://eospublic.cern.ch//eos/opendata/atlas/OutreachDatasets/2020-01-22"
+files = json.load(open(os.path.join(os.environ["ROOTSYS"], "tutorials/dataframe", "df107_SingleTopAnalysis.json")))
+processes = files.keys()
+df = {}
+xsecs = {}
+sumws = {}
+samples = []
+for p in processes:
+    for d in files[p]:
+        # Construct the dataframes
+        folder = d[0] # Folder name
+        sample = d[1] # Sample name
+        xsecs[sample] = d[2] # Cross-section
+        sumws[sample] = d[3] # Sum of weights
+        num_events = d[4] # Number of events
+        samples.append(sample)
+        df[sample] = ROOT.RDataFrame("mini", "{}/1lep/{}/{}.1lep.root".format(path, folder, sample))
+
+        # Scale down the datasets if requested
+        if not (args.t or lumi_scale == 1.0):
+            df[sample] = df[sample].Range(int(num_events * lumi_scale))
+
+# Select events for the analysis and make histograms of the top mass
+ROOT.gInterpreter.Declare("""
+using VecF_t = const ROOT::RVec<float>&;
+using VecI_t = const ROOT::RVec<int>&;
+int FindGoodLepton(VecI_t goodlep, VecI_t type, VecF_t pt, VecF_t eta, VecF_t phi, VecF_t e, VecF_t trackd0pv, VecF_t tracksigd0pv, VecF_t z0)
+{
+    int idx = -1;
+    for(auto i = 0; i < type.size(); i++) {
+        if(!goodlep[i]) continue;
+        if (type[i] == 11 && abs(eta[i]) < 2.47 && (abs(eta[i]) < 1.37 || abs(eta[i]) > 1.52) && abs(trackd0pv[i] / tracksigd0pv[i]) < 5) {
+            const ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+            if (abs(z0[i] * sin(p.Theta())) < 0.5) {
+                if (idx == -1) idx = i;
+                else return -1;
+            }
+        }
+        if (type[i] == 13 && abs(eta[i]) < 2.5 && abs(trackd0pv[i] / tracksigd0pv[i]) < 3) {
+            const ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+            if (abs(z0[i] * sin(p.Theta())) < 0.5) {
+                if (idx == -1) idx = i;
+                else return -1;
+            }
+        }
+    }
+    return idx;
+}
+
+float ComputeTransverseWMass(float lep_pt, float lep_eta, float lep_phi, float lep_e, float met_et, float met_phi)
+{
+    const ROOT::Math::PtEtaPhiEVector lep(lep_pt, lep_eta, lep_phi, lep_e);
+    const ROOT::Math::PtEtaPhiEVector met(met_et, 0, met_phi, met_et);
+    return sqrt(2 * lep.Pt() * met.Et() * (1 - cos(ROOT::Math::VectorUtil::DeltaPhi(lep, met))));
```

Comment:
Although I like working with four-vector in general, it's overkill in this case. You are using `lep_pt`, `met_et`, and literally a `phi_1 - phi_2` with range reduction. We clearly should offer this also with floats/doubles, instead of these vector monsters. Would be a good opportunity to add this to our library.

Reply:
Good catch! Absolutely right :)